### PR TITLE
Cleanup attribute bag usage in mixer/pkg/api.

### PR DESCRIPTION
**Pull request by geeknoid**
- Use snapshotting feature to fix bug where Check was not returning the correct set of
referenced attributes for quota results. It wasn't including attributes referenced from
the APA.

- Get rid of the compatibility bag goo from the gRPC server code. This is no longer
needed.

### CODE REVIEW

File: mixer/pkg/api/grpcServer.go [hunk 1]
```diff
@@ -122,42 +87,35 @@ func (s *grpcServer) Check(legacyCtx legacyContext.Context, req *mixerpb.CheckRe
...
 	if log.DebugEnabled() {
 		log.Debuga("Preprocess Check returned with: ", status.String(out))
 		log.Debug("Dispatching to main adapters after running processors")
-		log.Debuga("Attribute Bag: \n", mutableBag)
+		log.Debuga("Attribute Bag: \n", checkBag)
 		log.Debug("Dispatching Check")
 	}
 
-	cr, err := s.dispatcher.Check(legacyCtx, compatRespBag)
+	snap := protoBag.SnapshotReferencedAttributes()
```
Comment (ozevren) [inline mixer/pkg/api/grpcServer.go]: suggestion: rename snap to be more descriptive? i.e. apaSnapshot?

File: mixer/pkg/api/grpcServer.go [hunk 2]
```diff
@@ -202,11 +159,13 @@ func (s *grpcServer) Check(legacyCtx legacyContext.Context, req *mixerpb.CheckRe
 			}
 			var err error
 
-			qr, err = quota(legacyCtx, s.dispatcher, compatRespBag, qma)
+			// restore to the post-APA state
+			protoBag.ResetReferencedAttributes(snap)
```
Comment (ozevren) [inline mixer/pkg/api/grpcServer.go]: suggestion: ResetReferencedAttributes -> Restore?

File: mixer/pkg/api/grpcServer.go [hunk 3]
```diff
@@ -338,11 +292,11 @@ func (s *grpcServer) Report(legacyCtx legacyContext.Context, req *mixerpb.Report
 
 		span.LogFields(otlog.String("success", "finished Report for attribute bag "+string(i)))
 		span.Finish()
-		preprocResponseBag.Reset()
+		reportBag.Reset()
```
Comment (ozevren) [inline mixer/pkg/api/grpcServer.go]: A comment about why the reset of the reportBag is needed would be very helpful. It is hard to see that the bag is getting re-seeded with APA attributes.

File: mixer/pkg/api/grpcServer.go [hunk 4]
```diff
@@ -273,18 +231,17 @@ func (s *grpcServer) Report(legacyCtx legacyContext.Context, req *mixerpb.Report
 	}
 
 	protoBag := attribute.NewProtoBag(&req.Attributes[0], s.globalDict, s.globalWordList)
-	requestBag := attribute.GetMutableBag(protoBag)
-	compatReqBag := &compatBag{requestBag}
-	mutableBag := attribute.GetMutableBag(requestBag)
-	preprocResponseBag := attribute.GetMutableBag(nil)
+	accumBag := attribute.GetMutableBag(protoBag)
```
Comment (ozevren) [inline mixer/pkg/api/grpcServer.go]: A comment about why each bag is used would be great. It is hard to parse and grok what the function is doing, even more so with the loop and various uses of the bags. (Especially since the bag hierarchy is preserved across the iterations, but the individual bags get updated.)

**Comment by geeknoid** [comment]
Thanks for the feedback. I'll address all the naming & comment suggestions in my upcoming PR that will clean up this code considerably and get coverage up to snuff.
